Docs next: guide > cart and wishlist (#5468)
vuestorefront/vue-storefront: packages/core/docs/guide/cart-and-wishlist.md

# Cart and wishlist

## Introduction

> This document will guide you all the way to fully implement cart and wishlist modules, which will let you use these common features in your application in the best possible way. 

Reviewer: I think this is redundant, although let @NataliaTepluhina decide. Each section could start with a similar sentence and it doesn't provide any meaningful information IMHO. The sentence below explains what this document describes.

Author: changed

> UseCart and useWishlist are both composables which let you manage items by its properties and getters which you can import from the certain integration.  

Reviewer: ```suggestion
Customer's cart and wishlist can be managed using `useCart` and `useWishlist` composables respectively, provided by every integration. Data can be accessed using `cartGetters` and `wishlistGetters`. 
```

# Cart and wishlist

> ## Introduction

Author: removed

> ## Introduction

Reviewer 2: nope ;)

> ## Introduction

Author: sorry

> This document will guide you all the way to fully implement cart and wishlist modules, which will let you use these common features in your application in the best possible way. 

Author: removed

Customer's cart and wishlist can be managed using `useCart` and `useWishlist` composables respectively, provided by every integration. Data can be accessed using `cartGetters` and `wishlistGetters`. 

> ## Using cart

Author: removed

> ### loading cart in your application

Author: changed

The first thing that you need to do to start using the cart in your application is importing `useCart` from your ecommerce integration. Import `load`function which fetch the cart from a server and create a new one in your application if it doesn't already exist. Based on that, you can start using `cart` object, which structure depends on the ecommerce implementation you are using. 

```js
<script>
  import { useCart } from '{INTEGRATION}';
  import { onSSR } from '@vue-storefront/core';
  export default {
    setup() {
      const {
        cart,
        load
      } = useCart();
      onSSR(async () => {
        await load();
      });
      return {
        cart,
        load
      };
    }
  };   
</script> 
```

### Adding an item

> To add the product to the cart you need to apply `AddToCart` method in the following way:  

Author: fixed

```vue
<template>
  <CartComponent>
    <ProductComponent>
      <QuantityInput
        @input="updateQuantity(product, quantity)"
      />
      <RemoveButton        
        @click="removeFromCart(product)"
      />
    </ProductComponent>
    <TotalPriceComponent>
      {{totals}}
    </TotalPriceComponent>
    <TotalItemsComponent>
      {{totalItems}}
    </TotalItemsComponent>
  </CartComponent>
</template>   
<script>
  import { useCart, cartGetters } from '{INTEGRATION}';
  import { onSSR } from '@vue-storefront/core';
  export default {
    setup() {
      const {
        removeFromCart, 
        updateItemQty,
        loading
      } = useCart();

      const products = computed(() => cartGetters.getItems(cart.value));
      const totals = computed(() => cartGetters.getTotals(cart.value));
      const totalItems = computed(() => cartGetters.getTotalItems(cart.value));

      return {
        products,
        totals,
        totalItems,
        removeFromCart, 
        updateItemQty,
        loading
      };
    }
  };
</script>
```

> As you can see in the above example, we use getters to pass the values to the component: all products already in a cart, total price, and items in cart. It's the way to access `cart` object properties and, to make them work, you need to import `cartGetters` object and use a proper getter function as the computed value.    

Author: removed

```vue
<template>
  <ProductList>
    <ProductComponent>
      <AddedToCartIcon
        :isAddedToCart="isOnCart(product)"
      />
    </ProductComponent>
  </ProductList>
</template>    
<script>
  import { useCart } from '{INTEGRATION}';
  import { onSSR } from '@vue-storefront/core';
  export default {
    setup() {
      const {
        isOnCart, 
      } = useCart();

      return {
        isOnCart, 
      };
    }
  };
</script>
```

### Cleaning a cart

Implementing cleaning feature means that you should use `clear` function like in following example:

```vue
<template>
  <CartComponent>
    <ProductsList>
      <ProductComponent
        v-for="product in products"
      />
    <ProductsList>
    <ClearProductListButton
      @click="clear(cart.value)"
    />
  </CartComponent>
</template>   
<script>      
  import { useCart } from '{INTEGRATION}';
  export default {
    setup() {
      const {
        cart,
        clear, 
      } = useCart();
      const products = computed(() => cartGetters.getItems(cart.value));

      return {
        products,
        cart,
        clear 
      };
    }
  };
</script>
```

> ### Applying and removing coupons from a cart

Author: corrected

> Using Vue Storefront allows you to apply promotional coupons to your application and also use third-party software to handle that. Using `applyCoupon` function, you can pass the code for further handling.  

Author: changed

```vue
<template>
  <CartComponent>
    <CouponInput
      v-model="promoCode"
    />
    <ApplyCouponButton 
      @click="() => applyCoupon({ couponCode: promoCode })"
    >
      Use promo code
    </ApplyCouponButton>
    <RemoveCouponButton 
      @click="() => removeCoupon({ couponCode: promoCode })"
    >
      Remove promo code
    </RemoveCouponButton>
  </CartComponent>
</template>   
<script>      
  import { useCart } from '{INTEGRATION}';
  export default {
    setup() {
      const {
        cart,
        applyCoupon, 
      } = useCart();
      const products = computed(() => cartGetters.getItems(cart.value));
      const totals = computed(() => cartGetters.getTotals(cart.value));
      const totalItems = computed(() => cartGetters.getTotalItems(cart.value));

      return {
        products,
        totals,
        totalItems,
        cart,
        applyCoupon,
        removeCoupon
      };
    }
  };
</script>
```
 

> ## Using wishlist

Author: above applied also to wishlist

```vue
<template>
  ...
    <div>
      <div>
        <input
          @input="updateItemQty(product, quantity)"
        />
        <button        
          @click="removeItem(product)"
        >
          Remove product
        </button>
      </div>
      <span>
        {{totals}}
      </span>
      <span>
        {{totalItems}}
      </span>
    </div>
  ...
</template>   
<script>
  import { useCart, cartGetters } from '{INTEGRATION}';
  import { onSSR } from '@vue-storefront/core';
  export default {
    setup() {
      const {
        removeItem, 
        updateItemQty,
        loading
      } = useCart();

      const products = computed(() => cartGetters.getItems(cart.value));
      const totals = computed(() => cartGetters.getTotals(cart.value));
      const totalItems = computed(() => cartGetters.getTotalItems(cart.value));

      return {
        products,
        totals,
        totalItems,
        removeItem, 
        updateItemQty,
        loading
      };
    }
  };
</script>
```

## Checking if an item is in the cart

To check if a product is already in cart just pass it to `isInCart` method:

```vue
<template>
  ... 
    <ul>
      <li>

>         <icon

Author: forgot to remove

          :isAddedToCart="isInCart(product)"
        />
      </li>
    </ul>
  ...
</template>    
<script>
  import { useCart } from '{INTEGRATION}';
  import { onSSR } from '@vue-storefront/core';
  export default {
    setup() {
      const {
        isInCart, 
      } = useCart();

      return {
        isInCart, 
      };
    }
  };
</script>
```

## Removing all cart items at once

To clear cart items (not delete it) just use `clear` function.

```vue
<template>
  ...
    <div>
      <ul>
        <li
          v-for="product in products"
        >
          ...
        </li>
      <ul>
      <button
        @click="clear(cart.value)"
      >
        Clear cart
      </button>
    </div>
  ...
</template>   
<script>      
  import { useCart } from '{INTEGRATION}';
  export default {
    setup() {
      const {
        cart,
        clear, 
      } = useCart();
      const products = computed(() => cartGetters.getItems(cart.value));

      return {
        products,
        cart,
        clear 
      };
    }
  };
</script>
```

## Applying and removing discount coupons

> You can apply promotional coupons to your application with `applyCoupon` and remove with `removeCoupon` function:

Reviewer 2: ```suggestion
You can apply promotional coupons to your cart with `applyCoupon` and remove with `removeCoupon` function:
```

# Cart and wishlist

> > This page assumes you're familiar with Composition API and Composables. Read them first if you are new to these concepts.

Author: I'm not sure what is general approach. Because if we assume that user reads all docs one by one it's okay. But if you assume that someone just looks for certain solution that repetitive info makes sense.

## Introduction

This document will guide you all the way to fully implement cart and wishlist modules, which will let you use these common features in your application in the best possible way. 
Customer's cart and wishlist can be managed using `useCart` and `useWishlist` composables respectively, provided by every integration. Data can be accessed using `cartGetters` and `wishlistGetters`. 

## Using cart

### loading cart in your application

The first thing that you need to do to start using the cart in your application is importing `useCart` from your ecommerce integration. Import `load`function which fetch the cart from a server and create a new one in your application if it doesn't already exist. Based on that, you can start using `cart` object, which structure depends on the ecommerce implementation you are using. 

```js
<script>

>   import { useCart } from '@vue-storefront/{INTEGRATION}';

Author: ok

  import { onSSR } from '@vue-storefront/core';
  export default {
    setup() {
      const {
        cart,
        load
      } = useCart();
      onSSR(async () => {
        await load();
      });
      return {
        cart,
        load
      };
    }
  };   
</script> 
```

> ### basic usage scenarios:

Author: Yes you are right I'll change them.

  * adding an item

> The basic thing that a user can do in ecommerce application is adding an item to it. So having example application, we use `addToCart` function in order to allow user to add new product or additional items of product that is already added.   

Author: Ok, I'll replace them.

```vue
<template>
  <ProductList>
    <ProductComponent>
      <AddToCartButton
        @click="addToCart(product, parseInt(quantity))"
      />
    </ProductComponent>
  </ProductList>
</template>    
<script>     
  import { useCart } from '@vue-storefront/{INTEGRATION}';
  export default {
    setup() {
      const {
        addToCart,
        loading 
      } = useCart();

      return {
        addToCart,
        loading
      };
    }
  };
</script>
```

> After importing the `useCart`, you need to get proper function and pass the product and quantity of items. `loading` is the object which includes the loading state of the element.   

Author: removed

  export default {
    setup() {
      const {
        cart,
        load
      } = useCart();

      onSSR(async () => {
        await load();
      });

      return {
        cart,
        load
      };
    }
  };   
</script> 
```

## Adding an item to the cart

To add the product to the cart you can use `addItem` method:

```vue
<template>
  // ...
    <ul>
      <li
        v-for="(product, i) in products" :key="i"
      > 
        // ...
        <button
          @click="addItem({ product, quantity })"
        >
          Add to cart
        </button>
      </li>
    </ul>
  // ...
</template>    
<script>     
  import { computed } from '@vue/composition-api';
  import { useCart } from '{INTEGRATION}';
  export default {
    props: {
      products: {
        type: Array,
        required: true
      }
    },
    setup() {
      const {
        addItem,
      } = useCart();

      return {
        addItem,
      };
    }
  };
</script>
``` 

## Removing items and changing their quantity

To remove an item from the cart use `removeItem` function, and similarly to update quantity use `updateItemQty` function:

```vue
<template>
  // ...
    <div>
      <ul>
        <li

>           v-for="(product, i) in products" :key="i"

Reviewer 3: **nitpick**: could we use `product.id` as a key here? Using index is not optimal and we don't recommend it on Vue 3 docs. Total nitpick, please feel free to ignore 😅

        updateItemQty,
        loading
      } = useCart();

      const products = computed(() => cartGetters.getItems(cart.value);
      const totals = computed(() => cartGetters.getTotals(cart.value));
      const totalItems = computed(() => cartGetters.getTotalItems(cart.value));

      return {
        products,
        totals,
        totalItems,
        removeItem, 
        updateItemQty,
        loading
      };
    }
  };
</script>
```

## Checking if an item is in the cart

To check if a product is already in the cart, pass it to `isInCart` method:

```vue
<template>
  // ... 
    <ul>
      <li
         v-for="(product, i) in products" :key="i"
      >
        <div
          :isAddedToCart="isInCart({ product })"
        >
        </div>
      </li>
    </ul>
  // ...
</template>    
<script>
  import { computed } from '@vue/composition-api';
  import { useCart } from '{INTEGRATION}';

  export default {
    props: {
      products: {
        type: Array,
        required: true
      }
    },
    setup() {
      const {
        isInCart, 
      } = useCart();

      return {
        isInCart, 
      };
    }
  };
</script>
```

## Removing all cart items at once

> To clear cart items (not delete it) use `clear` function.

Reviewer 3: ```suggestion
To clear cart items (not delete it) use `clear` method.
```

```vue
<template>
  // ...
    <div>
      <ul>
        <li
          v-for="(product, i) in products" :key="i"
        >
          // ...
        </li>
      <ul>
      <button
        @click="clear"
      >
        Clear cart
      </button>
    </div>
  // ...
</template>   
<script>      
  import { computed } from '@vue/composition-api';
  import { useCart, cartGetters } from '{INTEGRATION}';

  export default {
    setup() {
      const {
        cart,
        clear, 
      } = useCart();
      const products = computed(() => cartGetters.getItems(cart.value));

      return {
        products,
        clear 
      };
    }
  };
</script>
```

## Applying and removing discount coupons

> You can apply promotional coupons to your cart with `applyCoupon` and remove with `removeCoupon` function:

Reviewer 3: ```suggestion
You can apply promotional coupons to your cart with `applyCoupon` and remove with `removeCoupon` method:
```

```vue
<script>
  import { useWishlist } from '{INTEGRATION}'
  import { onSSR } from '@vue-storefront/core'

  export default {
    setup () {
      const { 
        wishlist,      
        load,
      } = useWishlist()
      
      onSSR(async () => {
        await load()
      })

      return {
        wishlist,
        load      
      }
    }
  }
</script>
```

## Adding an item to the wishlist

To add the product to the wishlist you can use `addItem` method:

```vue
<template>
  // ...
    <ul>
      <li
        v-for="(product, i) in products" :key="i"
      > 
        // ...
        <button
          @click="addItem({ product, quantity })"
        >
          Add to wishlist
        </button>
      </li>
    </ul>
  // ...
</template>
<script>
  import { computed } from '@vue/composition-api';
  import { useWishlist } from '{INTEGRATION}'

  export default {
    props: {
      products: {
        type: Array,
        required: true
      }
    },
    setup (props) {
      const { addItem } = useWishlist();

      return {
        addItem,   
      }
    }
  }
</script>
```

## Removing an item from the wishlist
 

> To remove an item from the cart use `removeItem` function.

Reviewer 3: ```suggestion
To remove an item from the cart use `removeItem` method.
```

  import { useWishlist} from '{INTEGRATION}';
  export default {
    props: {
      products: {
        type: Array,
        required: true
      }
    },
    setup() {
      const {
        cart,
        isInWishlist, 
      } = useWishlist();

      return {
        isInWishlist, 
      };
    }
  };
</script>
```

## Removing all wishlist items at once

Cleaning the wishlist can be achieved by `clear` property.

```vue
<template>
  // ...
    <div>
      <ul>
        <li
          v-for="(product, i) in products" :key="i"
        >
          // ...
        </li>
      <ul>
      <button
        @click="clear()"
      >
        Clear wishlist
      </button>
    </div>
  // ...
</template>   
<script>      
  import { computed } from '@vue/composition-api';
  import { useWishlist, wishlistGetters } from '{INTEGRATION}';

  export default {
    setup() {
      const {
        wishlist,
        clear, 
      } = useCart();
      const products = computed(() => wishlistGetters.getItems(wishlist.value));

      return {
        wishlist,
        products,
        clear 
      };
    }
  };
</script>
```

## Summary

### Common usage example

> In the following examples, you can analyze how both composables are used in the simple use case. There are all above mentioned basic scenarios used in three main components: a product list, a cart, and a wishlist. It can be your basis for building real-life application implementation.    

Reviewer 3: I believe we could simplify wording here a bit. Also, it's not completely clear from the context which scenarios we mean.

```suggestion
Here is an example of how both composables can be used in real-life applications. We have three components: a product list, a cart, and a wishlist. In the examples, we cover the logic of adding/removing items from the wish list and cart, working with getters, and loading state
```

I might be missing something in the description 😅

```vue
<template>

>     <ul>

Reviewer 3: **suggestion**: could we add a comment here specifying it's a product list?
